File: src/profiles.ts
import AWS from "aws-sdk";
import log from "electron-log/main";
import slugify from "slugify";
import { config, UserConfig } from "./config.js";
import { writeAwsConfig } from "./aws-config.js";

const { SSO } = AWS;

const PREDEFINED_SHORT_NAMES: Record<string, string> = {
    AdministratorAccess: "admin",
    Billing: "billing",
    DatabaseAdministrator: "dba",
    DataScientist: "datasci",
    NetworkAdministrator: "netadmin",
    PowerUserAccess: "poweruser",
    SecurityAudit: "secaudit",
    SupportUser: "support",
    SystemAdministrator: "sysadmin",
    ViewOnlyAccess: "viewonly",
};

export async function refreshProfiles(): Promise<Profile[]> {
    log.info("[refreshProfiles] Refreshing profiles");

    const userConfig = config.get("userConfig") as UserConfig;
    const accessToken = config.get("accessToken") as string;

    const sso = new SSO({ region: userConfig.region });

    const accounts = await getAccounts(sso, accessToken);
    log.info("[refreshProfiles] Accounts: %s", JSON.stringify(accounts));

    const roles = ([] as AWS.SSO.RoleListType).concat.apply(
        [],
        await Promise.all(
            accounts.map((account) =>
                getAccountRoles(sso, accessToken, account.accountId!)
            )
        )
    );
    log.info("[refreshProfiles] Roles: %s", JSON.stringify(roles));

    const profiles = generateProfiles(userConfig, accounts, roles);
    await writeAwsConfig(profiles);
    return profiles;
}

async function getAccounts(
    sso: AWS.SSO,
    accessToken: string
): Promise<AWS.SSO.AccountListType> {
    let result: AWS.SSO.AccountListType = [];
    let nextToken: AWS.SSO.NextTokenType | undefined;

    do {
        const res: AWS.SSO.ListAccountsResponse = await sso
            .listAccounts({
                accessToken,
                nextToken,
            })
            .promise();
        result = result.concat(res.accountList || []);
        nextToken = res.nextToken;
    } while (nextToken);

    return result;
}

async function getAccountRoles(
    sso: AWS.SSO,
    accessToken: string,
    accountId: string
): Promise<AWS.SSO.RoleListType> {
    let result: AWS.SSO.RoleListType = [];
    let nextToken: AWS.SSO.NextTokenType | undefined;

    do {
        const res = await sso
            .listAccountRoles({
                accessToken,
                accountId,
                nextToken,
            })
            .promise();
        result = result.concat(res.roleList || []);
        nextToken = res.nextToken;
    } while (nextToken);

    return result;
}

export interface Profile {
    name: string;
    accountName: string;
    roleName: string;
    contents: {
        sso_start_url: string;
        sso_region: string;
        sso_account_id: string;
        sso_role_name: string;
        region: string;
        output: string;
    };
}

function generateProfiles(
    userConfig: UserConfig,
    accounts: AWS.SSO.AccountListType,
    roles: AWS.SSO.RoleListType
): Profile[] {
    const accountIdToName = new Map<string, string>(
        accounts.map((account) => [
            account.accountId!,
            shortAccountName(account.accountName!),
        ])
    );

    const accountIdToRegion = new Map<string, string>(
        accounts.map((account) => [
            account.accountId!,
            prefferedAccountRegion(account.accountName!) || userConfig.region,
        ])
    );

    return roles.map((role) => {
        const accountName = accountIdToName.get(role.accountId!);
        const roleName = shortPermissionSetName(role.roleName!);
        return {
            name: `${accountName}-${roleName}`,
            accountName,
            roleName,
            contents: {
                sso_start_url: userConfig.startUrl,
                sso_region: userConfig.region,
                sso_account_id: role.accountId!,
                sso_role_name: role.roleName!,
                region: accountIdToRegion.get(role.accountId!),
                output: "json",
            },
        } as Profile;
    });
}

function shortAccountName(name: string): string {
    const regex = /#([-_a-zA-Z0-9]+)/gm;
    const match = regex.exec(name);
    return match ? match[1] : slugify(name, { lower: true });
}

function prefferedAccountRegion(name: string): string | undefined {
    const regex = /@([a-zA-Z]+-[a-zA-Z]+-[0-9]+)/gm;
    const match = regex.exec(name);
    return match ? match[1] : undefined;
}

function shortPermissionSetName(name: string): string {
    return PREDEFINED_SHORT_NAMES[name] || slugify(name, { lower: true });
}

File: frost-core/src/profiles.rs
//! Generation of `~/.aws/config` profile names and contents.
//!
//! Ported from the original `src/profiles.ts`. The rules:
//!   * Account names may carry a `#short-name` tag and an `@region` tag.
//!   * Permission-set names are shortened via a known table, else slugified.
//!   * A profile is emitted per (account, role) pair, named `account-role`.

use crate::config::UserConfig;
use serde::{Deserialize, Serialize};

/// AWS-predefined permission sets mapped to their short forms.
fn predefined_short_name(name: &str) -> Option<&'static str> {
    Some(match name {
        "AdministratorAccess" => "admin",
        "Billing" => "billing",
        "DatabaseAdministrator" => "dba",
        "DataScientist" => "datasci",
        "NetworkAdministrator" => "netadmin",
        "PowerUserAccess" => "poweruser",
        "SecurityAudit" => "secaudit",
        "SupportUser" => "support",
        "SystemAdministrator" => "sysadmin",
        "ViewOnlyAccess" => "viewonly",
        _ => return None,
    })
}

/// An AWS account as returned by SSO `ListAccounts`.
#[derive(Debug, Clone)]
pub struct Account {
    pub account_id: String,
    pub account_name: String,
}

/// A role (permission set) available to the user in an account.
#[derive(Debug, Clone)]
pub struct Role {
    pub account_id: String,
    pub role_name: String,
}

/// The contents written under `[profile <name>]` in `~/.aws/config`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProfileContents {
    pub sso_start_url: String,
    pub sso_region: String,
    pub sso_account_id: String,
    pub sso_role_name: String,
    pub region: String,
    pub output: String,
}

/// A generated AWS profile.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Profile {
    pub name: String,
    pub account_name: String,
    pub role_name: String,
    pub contents: ProfileContents,
}

/// Extract the `#short-name` tag from an account name, or slugify the name.
fn short_account_name(name: &str) -> String {
    // First `#tag` of word characters / dashes wins.
    if let Some(m) = regex_first_group(name, r"#([-_a-zA-Z0-9]+)") {
        return m;
    }
    slug::slugify(name)
}

/// Extract a preferred region from an `@region` tag in the account name.
fn preferred_account_region(name: &str) -> Option<String> {
    regex_first_group(name, r"@([a-zA-Z]+-[a-zA-Z]+-[0-9]+)")
}

/// Shorten a permission-set name via the known table, else slugify it.
fn short_permission_set_name(name: &str) -> String {
    predefined_short_name(name)
        .map(str::to_owned)
        .unwrap_or_else(|| slug::slugify(name))
}

/// Run `pattern` against `text` and return the first capture group.
fn regex_first_group(text: &str, pattern: &str) -> Option<String> {
    let re = regex::Regex::new(pattern).expect("static regex");
    re.captures(text)
        .and_then(|c| c.get(1))
        .map(|m| m.as_str().to_owned())
}

/// Build the full set of profiles from accounts + roles.
pub fn generate_profiles(
    user_config: &UserConfig,
    accounts: &[Account],
    roles: &[Role],
) -> Vec<Profile> {
    roles
        .iter()
        .map(|role| {
            let account = accounts.iter().find(|a| a.account_id == role.account_id);
            let account_name = account
                .map(|a| short_account_name(&a.account_name))
                .unwrap_or_default();
            let region = account
                .and_then(|a| preferred_account_region(&a.account_name))
                .unwrap_or_else(|| user_config.region.clone());
            let role_name = short_permission_set_name(&role.role_name);

            Profile {
                name: format!("{account_name}-{role_name}"),
                account_name,
                role_name,
                contents: ProfileContents {
                    sso_start_url: user_config.start_url.clone(),
                    sso_region: user_config.region.clone(),
                    sso_account_id: role.account_id.clone(),
                    sso_role_name: role.role_name.clone(),
                    region,
                    output: "json".to_owned(),
                },
            }
        })
        .collect()
}
